File: app/src/main/java/xyz/rodit/snapmod/util/ConversationManager.java
package xyz.rodit.snapmod.util;

import android.text.TextUtils;

import java.io.File;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

import de.robv.android.xposed.XposedBridge;
import xyz.rodit.xposed.utils.StreamUtils;

public class ConversationManager {

    private final Set<String> conversations = new HashSet<>();
    private final File file;

    private long lastLoaded;

    public ConversationManager(File rootDir, String fileName) {
        file = new File(rootDir, fileName);
    }

    public Set<String> getEnabled() {
        if (file.lastModified() > lastLoaded) {
            load();
        }

        return conversations;
    }

    public void enable(String key) {
        if (!conversations.contains(key)) {
            conversations.add(key);
            save();
        }
    }

    public void disable(String key) {
        if (conversations.remove(key)) {
            save();
        }
    }

    public void toggle(String key) {
        if (isEnabled(key)) {
            disable(key);
        } else {
            enable(key);
        }
    }

    public boolean isEnabled(String key) {
        return getEnabled().contains(key);
    }

    private void load() {
        try {
            conversations.clear();
            if (file.exists()) {
                String[] keys = StreamUtils.readFile(file).split("\n");
                for (String key : keys) {
                    if (!TextUtils.isEmpty(key)) {
                        conversations.add(key);
                    }
                }
            }

            lastLoaded = System.currentTimeMillis();
        } catch (IOException e) {
            XposedBridge.log("Error loading conversation data.");
            XposedBridge.log(e);
        }
    }

    private void save() {
        try {
            StringBuilder keys = new StringBuilder();
            for (String key : conversations) {
                keys.append(key)
                        .append('\n');
            }
            StreamUtils.writeFile(file, keys.toString());
            lastLoaded = System.currentTimeMillis();
        } catch (IOException e) {
            XposedBridge.log("Error saving conversation data.");
            XposedBridge.log(e);
        }
    }
}

File: app/src/main/java/xyz/rodit/snapmod/util/ConversationManager.kt
package xyz.rodit.snapmod.util

import xyz.rodit.snapmod.logging.log
import java.io.File
import java.io.IOException

class ConversationManager(rootDir: File, fileName: String) {

    private val conversations: MutableSet<String> = HashSet()
    private val file: File = File(rootDir, fileName)
    private var lastLoaded: Long = 0

    val enabled: Set<String>
        get() {
            if (file.lastModified() > lastLoaded) {
                load()
            }

            return conversations
        }

    fun enable(key: String?) {
        key?.let {
            if (!conversations.contains(it)) {
                conversations.add(it)
                save()
            }
        }
    }

    fun disable(key: String?) {
        if (conversations.remove(key)) {
            save()
        }
    }

    fun toggle(key: String?) {
        if (isEnabled(key)) {
            disable(key)
        } else {
            enable(key)
        }
    }

    fun isEnabled(key: String?): Boolean {
        return enabled.contains(key)
    }

    private fun load() {
        try {
            conversations.clear()
            if (file.exists()) {
                file.readText().split("\n").filter { it.isNotEmpty() }.forEach(conversations::add)
            }

            lastLoaded = System.currentTimeMillis()
        } catch (e: IOException) {
            log.error("Error loading conversation data.", e)
        }
    }

    private fun save() {
        try {
            file.writeText(conversations.joinToString("\n"))
            lastLoaded = System.currentTimeMillis()
        } catch (e: IOException) {
            log.error("Error saving conversation data.", e)
        }
    }
}